src/function/table/arrow.cpp

```diff
@@ -0,0 +1,260 @@
+	ArrowSchema schema_root;
+	ArrowArray current_chunk_root;
+	idx_t chunk_idx = 0;
+	idx_t chunk_offset = 0;
+
+	void ReleaseArray() {
+		if (current_chunk_root.release) {
+			for (idx_t child_idx = 0; child_idx < (idx_t)current_chunk_root.n_children; child_idx++) {
+				auto &child = *current_chunk_root.children[child_idx];
+				if (child.release) {
+					child.release(&child);
+				}
+			}
+			current_chunk_root.release(&current_chunk_root);
+		}
+	}
+
+	void ReleaseSchema() {
+		if (schema_root.release) {
+			for (idx_t child_idx = 0; child_idx < (idx_t)schema_root.n_children; child_idx++) {
+				auto &child = *schema_root.children[child_idx];
+				if (child.release) {
+					child.release(&child);
+				}
+			}
+			schema_root.release(&schema_root);
+		}
+	}
+
+	~ArrowScanFunctionData() {
+		ReleaseSchema();
+		ReleaseArray();
+	}
+};
+
+static unique_ptr<FunctionData> arrow_scan_bind(ClientContext &context, vector<Value> &inputs,
+                                                unordered_map<string, Value> &named_parameters,
+                                                vector<LogicalType> &return_types, vector<string> &names) {
+
+	auto res = make_unique<ArrowScanFunctionData>();
+	auto &data = *res;
+
+	data.stream = (ArrowArrayStream *)inputs[0].GetValue<uintptr_t>();
+	if (!data.stream) {
+		throw InvalidInputException("arrow_scan: NULL pointer passed");
+	}
+
+	if (data.stream->get_schema(data.stream, &data.schema_root)) {
+		throw InvalidInputException("arrow_scan: get_schema failed(): %s",
+		                            string(data.stream->get_last_error(data.stream)));
+	}
+
+	if (!data.schema_root.release) {
+		throw InvalidInputException("arrow_scan: released schema passed");
+	}
+
+	if (data.schema_root.n_children < 1) {
+		throw InvalidInputException("arrow_scan: empty schema passed");
+	}
+
+	for (idx_t col_idx = 0; col_idx < (idx_t)data.schema_root.n_children; col_idx++) {
+		auto &schema = *data.schema_root.children[col_idx];
+		if (!schema.release) {
+			throw InvalidInputException("arrow_scan: released schema passed");
+		}
+		if (schema.dictionary) {
+			throw NotImplementedException("arrow_scan: dictionary vectors not supported yet");
+		}
+		auto format = string(schema.format);
+		if (format == "b") {
+			return_types.push_back(LogicalType::BOOLEAN);
+		} else if (format == "c") {
+			return_types.push_back(LogicalType::TINYINT);
+		} else if (format == "s") {
+			return_types.push_back(LogicalType::SMALLINT);
+		} else if (format == "i") {
+			return_types.push_back(LogicalType::INTEGER);
+		} else if (format == "l") {
+			return_types.push_back(LogicalType::BIGINT);
+		} else if (format == "f") {
+			return_types.push_back(LogicalType::FLOAT);
+		} else if (format == "g") {
+			return_types.push_back(LogicalType::DOUBLE);
+		} else if (format == "d:38,0") { // decimal128
+			return_types.push_back(LogicalType::HUGEINT);
+		} else if (format == "u") {
+			return_types.push_back(LogicalType::VARCHAR);
```

Comment:
are your VARCHAR required to be utf8? Curious how Arrow BINARY might be handled

Reply:
they are required to be UTF8 yes

Reply:
To add to that, we have a LogicalType::BLOB that can handle arbitrary binary blobs